lib/portlayer/vspc/vspc.go

```diff
@@ -0,0 +1,230 @@
+)
+const remoteConnReadDeadline = 1 * time.Second
+
+// VM is a struct that represents a VM
+type VM struct {
+	sync.Mutex
+	// this is the current connection between the vspc and the containerVM serial port
+	containerConn *telnet.Conn
+	// in case of vmotion, at some point we will have two telnet connections to the containerVM
+	// prevContainerConn is the connection of the source
+	// containerConn will be the connection to the destination
+	prevContainerConn *telnet.Conn
+	// inVmotion is a boolean denoting whether this VM is in a state of vmotion or not
+	inVmotion bool
+
+	// remoteConn is the remote-system connection
+	// this is the connection between the vspc and the attach-server
+	remoteConn net.Conn
+
+	vmUUID           string
+	vmotionStarted   chan chan struct{}
+	vmotionCompleted chan chan struct{}
+}
+
+// NewVM is the constructor of the VM
+func NewVM(tc *telnet.Conn) *VM {
+	return &VM{
+		containerConn:    tc,
+		inVmotion:        false,
+		vmotionStarted:   make(chan chan struct{}),
+		vmotionCompleted: make(chan chan struct{}),
+	}
+}
+
+// Vspc is all the vspc singletons
+type Vspc struct {
+	vmManagerMu *sync.Mutex
+	vmManager   map[string]*VM
+
+	*telnet.Server
+
+	attachSrvrAddr string
+	attachSrvrPort uint
+
+	doneCh chan bool
+}
+
+// NewVspc is the constructor
+func NewVspc(address string, port uint, attachSrvrAddr string, attachSrvrPort uint, ch chan bool) *Vspc {
+	vspc := &Vspc{
+		vmManagerMu: new(sync.Mutex),
+		vmManager:   make(map[string]*VM),
+
+		attachSrvrAddr: attachSrvrAddr,
+		attachSrvrPort: attachSrvrPort,
+
+		doneCh: ch,
+	}
+	hdlr := Handler{vspc}
+	opts := telnet.ServerOpts{
+		Addr:        fmt.Sprintf("%s:%d", address, port),
+		ServerOpts:  []byte{telnet.Binary, telnet.Sga, telnet.Echo},
+		ClientOpts:  []byte{telnet.Binary, telnet.Sga, VmwareExt},
+		DataHandler: hdlr.DataHdlr,
+		CmdHandler:  hdlr.CmdHdlr,
+	}
+	vspc.Server = telnet.NewServer(opts)
+	go vspc.monitorVMConnections()
+	return vspc
+}
+
+// getVM returns the VM struct from its uuid
+func (vspc *Vspc) getVM(uuid string) (*VM, bool) {
+	vspc.vmManagerMu.Lock()
+	defer vspc.vmManagerMu.Unlock()
+	if vm, ok := vspc.vmManager[uuid]; ok {
+		return vm, true
+	}
+	return nil, false
+}
+
+// addVM adds a VM to the map
+func (vspc *Vspc) addVM(uuid string, vm *VM) {
+	vspc.vmManagerMu.Lock()
+	defer vspc.vmManagerMu.Unlock()
+	vspc.vmManager[uuid] = vm
+}
+
+// relayReads reads from the AttachServer connection and relays the data to the telnet connection
+func (vspc *Vspc) relayReads(containervm *VM, conn net.Conn) {
+	vmotion := false
+	var tmpBuf bytes.Buffer
+	for {
+		select {
+		case ch := <-containervm.vmotionStarted:
+			vmotion = true
+			ch <- struct{}{}
+			log.Infof("vspc started to buffer data coming from the remote system")
+		case ch := <-containervm.vmotionCompleted:
+			vmotion = false
+			ch <- struct{}{}
+			log.Infof("vspc stopped buffering data coming from the remote system")
+		default:
+			b := make([]byte, 4096)
```

Comment:
we are copying back and forth here so why we are not replacing this while block with io.Copy?